core/templates/dev/head/services/contextual/UrlService.js

```diff
@@ -19,9 +19,13 @@
 
 oppia.factory('UrlService', ['$window', function($window) {
   return {
+    // This function is for testing purposes (to mock $window.location)
+    getCurrentUrl: function() {
+      return $window.location;
+    },
     getUrlParams: function() {
       var params = {};
-      var parts = $window.location.href.replace(
+      var parts = this.getCurrentUrl().href.replace(
```

Comment:
In this and other methods, don't forget to handle encoding/decoding of URI components. Add tests for these cases as well.

Reply:
Done